src/clang-cpp-frontend/clang_cpp_adjust_expr.cpp

```diff
@@ -282,3 +289,124 @@ void clang_cpp_adjust::align_se_function_call_return_type(
   if (return_type.id() != "constructor" && return_type.is_not_nil())
     expr.type() = return_type;
 }
+
+static bool dag_match_replace_helper(
+  exprt &to_search,
+  exprt &replacement,
+  const std::function<bool(exprt &)> &is_match)
+{
+  if (is_match(to_search))
+  {
+    to_search = replacement;
+    return true;
+  }
+  else
+  {
+    for (auto &op : to_search.operands())
+    {
+      if (dag_match_replace_helper(op, replacement, is_match))
+      {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+// Assumes that pattern and to_search are both DAGs.
+// Expects that there is no more than one match.
+// Returns true if a match was found and replaced.
+static bool dag_match_replace_one(
+  exprt &to_search,
+  exprt &replacement,
+  const std::function<bool(exprt &)> &is_match)
+{
+  bool result = dag_match_replace_helper(to_search, replacement, is_match);
+  if (result)
+  {
+    assert(!dag_match_replace_helper(to_search, replacement, is_match));
+  }
+  return result;
+}
+
+void clang_cpp_adjust::adjust_array_init_loop_expr(
+  exprt &lhs,
+  exprt &rhs,
+  exprt &new_expr)
+{
+  assert(rhs.statement() == "array_init_loop_expr");
+  assert(lhs.type().is_array());
+
+  exprt source_array_expr = rhs.op0();
+
+  typet array_element_type = source_array_expr.type().subtype();
+  typet array_element_ptr_type = pointer_typet(array_element_type);
+
+  exprt array_init = side_effect_exprt("assign", lhs.type());
+  array_init.copy_to_operands(lhs, rhs);
+  symbolt source_array_symbol = tmp_symbol.new_symbol(
+    context, array_element_ptr_type, "array_init_loop_expr_tmp");
+  context.add(source_array_symbol);
+  source_array_symbol.lvalue = true;
+  source_array_symbol.value = source_array_expr;
```

Comment:
The .new_symbol() call already registers the symbol with the context (hence it returns a reference), no need to add it again. In fact, it should be accessed by reference, otherwise you're modifying a copy outside of the context with the last 2 statements.